### ISSUE TITLE
[MNG-6254] Clarification on <versions> sorting in Repository Metadata

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#7616]
**[Markus Karg](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mkarg)** opened **[MNG-6254](https://issues.apache.org/jira/browse/MNG-6254?redirect=false)** and commented

The document http://maven.apache.org/ref/3.3.9/maven-repository-metadata/repository-metadata.html explains that \<versions> does contain a _List_\<Version>, hence is _sorted_.

When pulling _real _metadata from Maven Central, e. g. http://repo.maven.apache.org/maven2/javax/ws/rs/jsr311-api/maven-metadata.xml, it is apparent that the content of \<version> is _not_ sorted (in this example, 1.1-ea is listed _after_ 1.1, which is a violation of the sorting, because -ea must be listed _before_ 1.1). So effectively Maven Central devliers _unsorted_ versions.

Also, the effectively listed XSD http://maven.apache.org/xsd/metadata-1.1.0.xsd is nonexistent, so cannot be looked but whether Maven Central is wrong, or whether actually the specification should say it must be Set\<Version> instead of List\<Version>.

To sum up, there should be a clarification whether \<version> must be sorted according to the Maven Version Numbering Schema, whether it must be Set vs List, and whether the missing XSD should be uploaded.

---

**Affects:** 3.5.0

**Issue Links:**
- [MNG-7000](https://issues.apache.org/jira/browse/MNG-7000) metadata.mdo contains invalid link to schema

**Comment by jira-importer** [comment]
**[Manfred Moser](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=simpligility)** commented

Afaik it is sorted by time of deployment... each deployment (independent of version) appends as new entry. So its sorted .. just not the way you might want it.

**Comment by jira-importer** [comment]
**[Markus Karg](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mkarg)** commented

Actually the aim of this bug report is that make clear what the specification wants, neither what you "AFAIK" nor what I want. It is to be clarified and clearly documented once.

**Comment by jira-importer** [comment]
**[Manfred Moser](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=simpligility)** commented

I do not think it is realistic to reprocess all existing metadata files based on some new desired "correct" behavior. Instead I propose to document how it actually works....

**Comment by jira-importer** [comment]
**[Markus Karg](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mkarg)** commented

Just as I said, all I propose is to clarify and document, not to rework anything.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

The incorrect location has been fixed by another ticket. The reaon by I am closing this is that your assumption is incorrect. `java.util.List` says: An ordered collection (also known as a sequence).
No sentence about sorted. No assumption can be made about the sorting or specific order of elements. It is an implementation detail. It is imprerative to sort manually, in fact, there is no reliable sorting order because your artifact version class does not cover all cases.
